File: c++/diffusion.cpp
/*
Ervin Pangilinan
CSC 330: Organization of Programming Languages - Fall 2022
Project 2: Simplified 3D Diffusion Model
Implementation in C++
*/

/*
To compile: c++ diffusion.cpp -O2
To execute: ./[binary file name] [number of subdivisions] ["partition" (leave blank if none)]
*/

using namespace std;
#include <iostream>
#include <algorithm>
#include <cmath>
#include <vector>
typedef vector<vector<vector<double>>> Cube;
typedef vector<vector<double>> Matrix;
typedef vector<vector<vector<int>>> BitMask;
typedef vector<vector<int>> CrossSection;

// Overloaded & dependent on if a flag has been set to put up a partition in the room.
void diffusion(Cube &room, BitMask &BitMask, const int &maxSize, const double &DTerm);     // Executes if there is a partition.
void diffusion(Cube &room, const int &maxSize, const double &DTerm);                       // Executes if there isn't.
void setBitMask(Cube &BitMask, const int &maxSize);

void diffusion(Cube &room, const int &maxSize, const double &DTerm) {
    // PRE: 3D array for the room and number of subdivisions have been initialized.
    // POST: Simulates the diffusion of molecules and writes into the 3D array of the room.

    double change;
    for (int i = 0; i < maxSize; i++) {
        for (int j = 0; j < maxSize; j++) {
            for (int k = 0; k < maxSize; k++) {
                for (int l = 0; l < maxSize; l++) {
                    for (int m = 0; m < maxSize; m++) {
                        for (int n = 0; n < maxSize; n++) {
                            if (((i == l) && (j == m) && (k == n + 1)) ||  // move down
                                ((i == l) && (j == m) && (k == n - 1)) ||  // move up
                                ((i == l) && (j == m + 1) && (k == n)) ||  // move right
                                ((i == l) && (j == m - 1) && (k == n)) ||  // move left
                                ((i == l + 1) && (j == m) && (k == n)) ||  // move forward
                                ((i == l - 1) && (j == m) && (k == n))) {  // move backwards

                                change = (room[i][j][k] - room[l][m][n]) * DTerm;
                                room[i][j][k] = room[i][j][k] - change;
                                room[l][m][n] = room[l][m][n] + change;
                            }
                        }
                    }
                }
            }
        }
    }  
}

void diffusion(Cube &room, BitMask &BitMask, const int &maxSize, const double &DTerm) {
    // PRE: 3D array for the room, BitMask, and number of subdivisions have been initialized.
    // POST: Uses the BitMask to simulate diffusion with partition in the room. 

    double change;
    for (int i = 0; i < maxSize; i++) {
        for (int j = 0; j < maxSize; j++) {
            for (int k = 0; k < maxSize; k++) {
                for (int l = 0; l < maxSize; l++) {
                    for (int m = 0; m < maxSize; m++) {
                        for (int n = 0; n < maxSize; n++) {
                            if ((((i == l) && (j == m) && (k == n + 1)) ||      // move down
                                ((i == l) && (j == m) && (k == n - 1)) ||       // move up
                                ((i == l) && (j == m + 1) && (k == n)) ||       // move right
                                ((i == l) && (j == m - 1) && (k == n)) ||       // move left
                                ((i == l + 1) && (j == m) && (k == n)) ||       // move forward
                                ((i == l - 1) && (j == m) && (k == n))) &&      // move backwards
                                (BitMask[i + 1][j + 1][k + 1] == 0 && BitMask[l + 1][m + 1][n + 1] == 0)) {  
                                change = (room[i][j][k] - room[l][m][n]) * DTerm;
                                room[i][j][k] = room[i][j][k] - change;
                                room[l][m][n] = room[l][m][n] + change;
                            }
                        }
                    }
                }
            }
        }
    }
}

void setBitMask(BitMask &BitMask, const int &maxSize) {
    // PRE: BitMask and number of subdivisions have been initialized. 
    // POST: Adjusts the the BitMask by putting in 1 in indices where the partition is present or is a wall.

    // Set up the 75% partition.
    for (int j = floor((maxSize / 4) + 1); j < maxSize + 2; j++) {
        for (int k = 0; k < maxSize + 2; k++) {
            BitMask[maxSize / 2][j][k] = 1;
        }
    }

    // Set up the front-facing wall.
    for (int j = 0; j < maxSize + 2; j++) {
        for (int k = 0; k < maxSize + 2; k++) {
            BitMask[0][j][k] = 1;
        }
    }

    // Set up the back-facing wall. 
    for (int j = 0; j < maxSize + 2; j++) {
        for (int k = 0; k < maxSize + 2; k++) {
            BitMask[maxSize + 1][j][k] = 1;
        }
    }

    // Set up the room's ceiling.
    for (int i = 0; i < maxSize + 2; i++) {
        for (int k = 0; k < maxSize + 2; k++) {
            BitMask[i][0][k] = 1;
        }
    }

    // Set up the room's floor. 
    for (int i = 0; i < maxSize + 2; i++) {
        for (int k = 0; k < maxSize + 2; k++) {
            BitMask[i][maxSize + 1][k] = 1;
        }
    }

    // Set up the left-facing wall.
    for (int i = 0; i < maxSize + 2; i++) {
        for (int j = 0; j < maxSize + 2; j++) {
            BitMask[i][j][0] = 1;
        }
    }

    // Set up the right-facing wall. 
    for (int i = 0; i < maxSize + 2; i++) {
        for (int j = 0; j < maxSize + 2; j++) {
            BitMask[i][j][maxSize + 1] = 1;
        }
    }
}

// Basing implementation on Java sample code from class.
int main(int argc, char *argv[]) {
    // PRE: Flags are set to determine room dimension and if there's a partition in the room.
    // POST: Outputs the ratio of concentration and time taken to equilibrate the room. 

    // Number of subdivisions per dimension is determined by command-line argument. Defaults to 10 if there's no argument. 
    const int maxSize = stoi(argv[1]) > 0 ? stoi(argv[1]) : 10;

    // Specify if there is a flag to set up partition in the room.
    string arg2 = argv[argc - 1];
    bool partitionPresent = arg2 == "partition" ? true : false;
    
    // Attempt at dynamic allocation for scalability.
    Cube room(maxSize, Matrix(maxSize, vector<double> (maxSize, 0))); 
    BitMask mask(maxSize + 2, CrossSection(maxSize + 2, vector<int> (maxSize + 2, 0)));
    if (partitionPresent) setBitMask(mask, maxSize);

    double diffusionCoefficient = 0.175;
    double roomDimension = 5;                                                   // In meters.
    double speedOfGasMolecules = 250.0;                                         // Based on 100 g/mol gas at RT.
    double timestep = (roomDimension / speedOfGasMolecules) / maxSize;          // h in seconds
    double distanceBetweenBlocks = roomDimension / maxSize;
    double DTerm = diffusionCoefficient * timestep / (distanceBetweenBlocks * distanceBetweenBlocks);
    double time = 0;
    double ratio = 0;

    // Initialize first cell so it first contains the molecules.
    room[0][0][0] = 1.0e21;

    do {
        if (partitionPresent) {
            diffusion(room, mask, maxSize, DTerm);
        } else {
            diffusion(room, maxSize, DTerm);
        }

        time += timestep;
        double sumval = 0.0;
        double maxval = room[0][0][0];
        double minval = room[0][0][0];

        for (int i = 0; i < maxSize; i++) {
            for (int j = 0; j < maxSize; j++) {
                for (int k = 0; k < maxSize; k++) {
                  if (room[i][j][k] != 0) {
                    maxval = max(room[i][j][k], maxval);
                    minval = min(room[i][j][k], minval);
                    sumval += room[i][j][k];          
                  }   
                }
            }
        }

        ratio = minval / maxval;
        std::cout << time << " \t " << room[0][0][0] <<  " \t ";
        std::cout << room[maxSize - 1][0][0] << " \t ";
        std::cout << room[maxSize - 1][maxSize - 1][0] << " \t ";
        std::cout << room[maxSize - 1][maxSize - 1][maxSize - 1] << " \t ";
        std::cout << ratio << endl;

    } while (ratio < 0.99);

    std::cout << "\nBox equilibrated in " << time << " seconds of simulated time." << endl;
    return 0;
}

File: rust/diffusion.rs
/*
Ervin Pangilinan
CSC 330: Organization of Programming Languages - Fall 2022
Project 2: Simplified 3D Diffusion Model
Implementation in Rust
*/

use std::env;

fn main() {
    // PRE: Command line arguments are passed to determine subdivisions and partition in room. 
    // POST: Outputs the ratio of concentration and time taken to equilibrate the room. 

    // Contains command-line flags to be parsed. 
    let args: Vec<String> = env::args().collect();      
    let max_size: i64 = args[1].parse().unwrap();
    let partition_present: bool = if args.len() - 1 >= 1 && args[args.len() - 1] == "partition" {true} else {false};

    // Pre-defined values. 
    const DIFFUSION_COEFFICIENT: f64 = 0.175;             
    const ROOM_DIMENSION: f64 = 5.0;                                                      // In meters. 
    const SPEED_OF_GAS_MOLECULES: f64 = 250.0;                                            // Based on 100 g/mol
    let timestep: f64 = (ROOM_DIMENSION / SPEED_OF_GAS_MOLECULES) / (max_size as f64);    // h in seconds
    let distance_between_blocks: f64 = ROOM_DIMENSION / (max_size as f64);
    let d_term: f64 = (DIFFUSION_COEFFICIENT * timestep) / (distance_between_blocks * distance_between_blocks);

    // Mutable variables to be used. 
    let mut time: f64 = 0.0;
    let mut ratio: f64 = 0.0;
    let mut room = vec![vec![vec![0.0; max_size as usize]; max_size as usize]; max_size as usize];
    let mut mask = vec![vec![vec![0; (max_size + 2) as usize]; (max_size + 2) as usize]; (max_size + 2) as usize];
    room[0][0][0] = 1e+21_f64;

    if partition_present {
        // Set up the 75% partition. 
        for j in (((max_size / 4) + 1) as f64).floor() as usize..(max_size + 2) as usize {
            for k in 0..(max_size + 2) as usize {
                mask[((max_size / 2) as f64).floor() as usize][j][k] = 1;
            }
        }

        // Set up the front-facing wall. 
        for j in 0..(max_size + 2) as usize {
            for k in 0..(max_size + 2) as usize {
                mask[0][j][k] = 1;
            }
        }

        // Set up the back-facing wall.
        for j in 0..(max_size + 2) as usize {
            for k in 0..(max_size + 2) as usize {
                mask[(max_size + 1) as usize][j][k] = 1;
            }
        }

        // Set up the room's ceiling.
        for i in 0..(max_size + 2) as usize {
            for k in 0..(max_size + 2) as usize {
                mask[i][0][k] = 1;
            }
        }

        // Set up the room's floor. 
        for i in 0..(max_size + 2) as usize {
            for k in 0..(max_size + 2) as usize {
                mask[i][(max_size + 1) as usize][k] = 1;
            }
        }

        // Set up the left-facing wall.
        for i in 0..(max_size + 2) as usize {
            for j in 0..(max_size + 2) as usize {
                mask[i][j][0] = 1;
            }
        }

        // Set up the right-facing wall. 
        for i in 0..(max_size + 2) as usize {
            for j in 0..(max_size + 2) as usize {
                mask[i][j][(max_size + 1) as usize] = 1;
            }
        }

    }

    while ratio < 0.99 {
        for i in 0..max_size {
            for j in 0..max_size {
                for k in 0..max_size {
                    for l in 0..max_size {
                        for m in 0..max_size {
                            for n in 0..max_size {
                                if  (i == l && j == m && k == n + 1) ||                     // move down
                                    (i == l && j == m && k == n - 1) ||                     // move up
                                    (i == l && j == m + 1 && k == n) ||                     // move right
                                    (i == l && j == m - 1 && k == n) ||                     // move left
                                    (i == l + 1 && j == m && k == n) ||                     // move forward
                                    (i == l - 1 && j == m && k == n) {                      // move backwards
                                    if partition_present {
                                        if mask[(i + 1) as usize][(j + 1) as usize][(k + 1) as usize] == 0 &&
                                        mask[(l + 1) as usize][(j + 1) as usize][(k + 1) as usize] == 0 {

                                            let change = (room[i as usize][j as usize][k as usize] 
                                                - room[l as usize][m as usize][n as usize]) * d_term;
                                            room[i as usize][j as usize][k as usize] -= change;
                                            room[l as usize][m as usize][n as usize] += change;
                                        }
                                    } else {
                                        let change = (room[i as usize][j as usize][k as usize] 
                                            - room[l as usize][m as usize][n as usize]) * d_term;
                                        room[i as usize][j as usize][k as usize] -= change;
                                        room[l as usize][m as usize][n as usize] += change;
                                    }   
                                }
                            }
                        }
                    }
                }
            }
        }

        time += timestep;
        let mut _sumval: f64 = 0.0;
        let mut minval: f64 = room[0][0][0];
        let mut maxval: f64 = room[0][0][0];
    
        for i in 0..max_size as usize {
            for j in 0..max_size as usize {
                for k in 0..max_size as usize {
                    if room[i][j][k] != 0.0 {
                        maxval = maxval.max(room[i][j][k]);
                        minval = minval.min(room[i][j][k]);
                        _sumval += room[i][j][k];
                    }  
                }
            }
        } 
    
        ratio = minval / maxval;
        println!("{:.3} \x09 {:.5e} \x09 {:.5e} \x09 {:.5e} \x09 {:.5e} \x09 {:.5e}", 
            time, 
            room[0][0][0], 
            room[max_size as usize - 1][0][0], 
            room[max_size as usize - 1][max_size as usize - 1][0], 
            room[max_size as usize - 1][max_size as usize - 1][max_size as usize - 1], ratio);
    }

    println!("\x0ABox equilibrated in {:.3} seconds of simulated time.", time);
}
